[Feat] Game socket & API

yeslee-v opened the pull request:
- create socket: `game-join-socket`, ``game-leave-socket`
- create API: gameRoomList, gameParticipantList
- **NOT YET position** in game.gateway

yeslee-v committed:
[Feat] craete initial socket: join game room

- create dto, controller, service
- create method(createRoom) in repository

yeslee-v committed:
[Merge] Merge branch 'develop' of https://github.com/Maserati-Quattroporte/maseraT_pong_game into feature/create_game_join_socket

- NOT YET join-game-room

yeslee-v committed:
Merge branch 'feature/create_game_join_socket' into develop

yeslee-v committed:
[Feat] game-participant API

- get game participant list, specific game participant info
- NOT YET game.gateway
  - set title
  - set Position

yeslee-v committed:
Merge branch 'develop' of https://github.com/Maserati-Quattroporte/maseraT_pong_game into develop

yeslee-v committed:
Merge branch 'develop' of https://github.com/Maserati-Quattroporte/maseraT_pong_game into develop

yeslee-v committed:
[Feat] create socket

- game-leave-socket
- deleteRoom method in repository

yeslee-v committed:
Merge branch 'develop' of https://github.com/Maserati-Quattroporte/maseraT_pong_game into develop

hsw7703 commented on backend/src/game/dto/game-participant.dto.ts:
```diff
@@ -0,0 +1,9 @@
+  gameRoomId: number;
+  title: string;
+  userId: number;
+  userNickname: string;
```
userNickname 말고 그냥 nickname으로 하는게 더 좋지 않을까요? 다른데선 다 nickname으로 보내고 있어서..

yeslee-v replied:
소켓 문서에 나와있는 그대로 입력했더니만..ㅋㅋ 접수하겠습니다

hsw7703 commented on backend/src/game/dto/game-room.dto.ts:
```diff
@@ -0,0 +1,12 @@
+  gameRoodId: number;
+  title: string;
+  userId: number;
+  userNickname: string;
```
여기도 그냥 nickname으로

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,103 @@ export class GameGateway {
+      let gameParticipant: GameParticipant;
+
+      // participant id에 따라 position 결정
+      if (gameParticipant.id === 1) {
```
gameParticipant에 아무 값도 없는데 id를 어떻게 비교하시죠? 선언만 해 놓으시고 아무 값도 안 집어 넣으셨는데

yeslee-v replied:
번외로, id로 비교하는게 맞는지 몰라서 조건만 넣은거였는데 추가를 안했네요. 전에 말해주신 position 관련해서 이해를 잘 못했어서요.

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,103 @@ export class GameGateway {
+      );
+    }
+
+    if (!this.joinGameRoom(gameJoinDto, user)) return;
```
async 걸려있는 함수 await 없이 쓰시면 pending 넘어 올 거 같아서 이 부분에 await 걸어주셔야 할거 같아요. 한 번 확인해 주세요

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,103 @@ export class GameGateway {
+      .in(gameParticipantDto.title)
+      .emit("game-room-join", gameParticipantDto);
+
+    socket.join(gameParticipantDto.title);
```
저희 title 중복 되서 채팅방은 chat-id, 게임방은 game-id 이렇게 넣기로 했었는데 이 부분 수정해주세요.
socket.join("game-" + gameParticipantDto.gameRoomId) 이런 식으로

yeslee-v replied:
title이 중복된다는 것을 이해 못했어요

yeslee-v replied:
방 제목을 애초에 입력받아서 title에 넣는게 아니었나요?

jeongwle replied:
방제목 자체가 중복 검사를 따로 하지 않기 떄문에 방1 이라는 title의 방을 아주 많이 생성할 수 있습니다. 그래서 구분하기 위해 room의 pk인 id를 넣어서 하자는 말인 것 같습니다.

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,103 @@ export class GameGateway {
+  ): Promise<void> {
+    const user: User = await this.userRepository.findOne(gameLeaveDto.userId);
+
+    socket.leave(gameLeaveDto.title);
```
join 하는 room이름 바꾸시면서 여기도 같이 바꿔주세요

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,103 @@ export class GameGateway {
+    this.server.in(gameLeaveDto.title).emit("game-room-leave"), gameLeaveDto);
+
+    const participant: GameParticipant =
+      await this.gameParticipantRepository.findOne(gameLeaveDto.userId);
```
findOne에 where 안 넣어서 find하시면 participant pk로 find하는데 pk랑 userid랑은 다른 값이여서 원하는 값이 안 나오겠네요.
그리고 위에서 유저 delete하셔서 찾으시려고 하는 값도 없을 거 같구요

hsw7703 commented on backend/src/game/game.service.ts:
```diff
@@ -0,0 +1,33 @@
+@Injectable()
+export class GameService {
+  constructor(
+    private gameJoinRepository: GameRoomRepository,
```
이 부분은 이름 gameJoinRepository 보단 gameRoomRepository가 더 좋을 거 같아요

hsw7703 commented on backend/src/game/game.service.ts:
```diff
@@ -0,0 +1,33 @@
+    const gameParticipants: GameParticipant[] =
+      await this.gameParticipantRepository.find({
+        where: { gameRoom: gameRoomId },
+        relations: ["User"],
```
User -> user 로 바꾸셔야해요

hsw7703 commented on backend/src/game/repository/game-room.repository.ts:
```diff
@@ -1,5 +1,27 @@
+
+    const gameRoom: GameRoom = this.create({
+      title,
+    });
```
isStart, isLadder 둘 다 default 값 없어서 이렇게 하시면 error 뜰 거 같은데 확인해 보셨나요?

hsw7703 reviewed (changes requested):
고쳐야 할 부분들 error 뜰 거 같은 부분들 comment 남겨 놨습니다. 확인후에 고쳐서 다시 pr 해주세요

yeslee-v committed:
[Fix] bugs

- change userNickname to nickname
- add game participant in game-room-join
- create variable for await method
- redeclare title for duplicated name
- add 'where' condition in findOne option
- redefine gameRoomRepository in service
- modify user at relations in service
- initialize isStart, isLadder

yeslee-v committed:
Merge branch 'develop' of https://github.com/Maserati-Quattroporte/maseraT_pong_game into develop

yeslee-v commented:
position 저렇게 짜는거 맞는지요.. 흠...... 헷갈리네요.
그리고 request로 받는 title과 직접 넣어주는 초기의 title과는 다른 것 맞죠?

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,132 @@ export class GameGateway {
+    gameJoinDto: GameJoinDto,
+    user: User,
+  ): Promise<boolean> {
+    const gameRoom: GameRoom = await this.gameRoomRepository.findOne({
```
pk로 find 할 때는 where문 안 써도 되요~

yeslee-v replied:
ㅎ 이제 다 까먹는군요

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,132 @@ export class GameGateway {
+      },
+    });
+
+    // 게임 첫 참여자는 무조건 leftUser인지?
```
넵!

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,132 @@ export class GameGateway {
+          gameRoom,
+        });
+
+      // participant id에 따라 position 결정
```
gameParticipants = this.gameParticipantRepository.find({
  where: {
    gameRoom: // 게임 룸 조회
  },
  order: {
    gamePosition: 1
  }
})
이렇게 조회하시면 position 오름차순 정렬로 나와서 [0], [1] user만 확인하시면 left, right중 어디 비어있는지 체크 가능할 거 같아요!

yeslee-v replied:
흐어 gamePosition을 정해놓고 그걸 체크할 생각을 안했군요..ㅡㅡ

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,132 @@ export class GameGateway {
+  ): Promise<void> {
+    const user: User = await this.userRepository.findOne({
+      where: {
+        id: gameLeaveDto.userId,
```
여기도 id로 조회할 땐 where 안 쓰셔도 되요

hsw7703 reviewed (changes requested):
조금만 더 수정 하시면 될 거 같아요

yeslee-v committed:
Merge branch 'develop' of https://github.com/Maserati-Quattroporte/maseraT_pong_game into develop

yeslee-v committed:
[Fix] bugs

- delete where condition when compare ids
- modify rules: set position if not exists position

yeslee-v committed:
[Fix] position in joinGameRoom

- save when do not exist gameRoomId
- set position when do not exist
- modify return type in joinGameRoom

yeslee-v committed:
[Modify] TRAN-10 username: maserati

yeslee-v committed:
[Fix] TRAN-52 TRAN-53 game gateway

- set return type in handleJoinGameRoom
- remove comments
- remove gameParticipantDto in handleGameRoomLeave

yeslee-v committed:
Merge branch 'develop' of https://github.com/Maserati-Quattroporte/maseraT_pong_game into develop

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,114 @@ export class GameGateway {
+
+    const existLeftUser = this.gameParticipantRepository.findOne({
+      position: GamePosition.leftUser,
+    });
```
game room id도 같이 찾아야되요!

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,114 @@ export class GameGateway {
+
+    const existSpectator = this.gameParticipantRepository.findOne({
+      position: GamePosition.spectator,
+    });
```
없어도 되요

yeslee-v replied:
default position으로 spectator 설정했습니다

hsw7703 commented on backend/src/game/game.gateway.ts:
```diff
@@ -42,6 +45,114 @@ export class GameGateway {
+        user,
+        gameRoom,
+      });
+    } else if (!existSpectator) {
```
left user, right user 둘다 있으면 무조건 spectator니까 else if -> else 쓰시면 될 거 같아요

hsw7703 reviewed (changes requested):
다시!! 고쳐주세요

yeslee-v committed:
[Fix] set default position(spectator)

- refactor code: set position when do not exist

hsw7703 reviewed (approved):
드디어!!! 4번만에? 5번만에? 통과 했네요!!

yeslee-v commented:
db 연결 안됐던 것도 해결되서 너무 좋네요ㅠㅠ
